logstash-core/src/main/java/org/logstash/secret/store/backend/JavaKeyStore.java

```diff
@@ -0,0 +1,209 @@
+                    byte[] base64 = SecretStoreUtil.base64Encode(LOGSTASH_MARKER.getBytes(StandardCharsets.UTF_8));
+                    SecretKey secretKey = factory.generateSecret(new PBEKeySpec(SecretStoreUtil.asciiBytesToChar(base64)));
+                    keyStore.setEntry(logstashMarker.toExternalForm(), new KeyStore.SecretKeyEntry(secretKey), protectionParameter);
+                    keyStore.store(os, this.keyStorePass);
+                    PosixFileAttributeView attrs = Files.getFileAttributeView(keyStorePath, PosixFileAttributeView.class);
+                    if (attrs != null) {
+                        //the directory umask applies when creating the file, so re-apply permissions here
+                        attrs.setPermissions(PosixFilePermissions.fromString(keyStorePermissions));
+                    }
+                    LOGGER.info("Keystore created at {}", keyStorePath.toAbsolutePath());
+                } catch (Exception e) {
+                    throw new SecretStoreException.CreateException("Failed to create Logstash keystore.", e);
+                }
+            }
+        } catch (SecretStoreException sse) {
+            throw sse;
+        } catch (Exception e) { //should never happen
+            throw new SecretStoreException("Error while trying to create or load the Logstash keystore", e);
+        }
+    }
+
+    @Override
+    public Collection<SecretIdentifier> list() {
+        Set<SecretIdentifier> identifiers = new HashSet<>();
+        try {
+            readLock.lock();
+            Enumeration<String> aliases = keyStore.aliases();
+            while (aliases.hasMoreElements()) {
+                String alias = aliases.nextElement();
+                identifiers.add(SecretIdentifier.fromExternalForm(alias));
+            }
+        } catch (KeyStoreException e) {
+            throw new SecretStoreException.ListException(e);
+        } finally {
+            readLock.unlock();
+        }
+        return identifiers;
+    }
+
+    @Override
+    public void persistSecret(SecretIdentifier identifier, byte[] secret) {
+        try {
+            writeLock.lock();
+            SecretKeyFactory factory = SecretKeyFactory.getInstance("PBE");
+            //PBEKey requires an ascii password, so base64 encode it
+            byte[] base64 = SecretStoreUtil.base64Encode(secret);
+            PBEKeySpec passwordBasedKeySpec = new PBEKeySpec(SecretStoreUtil.asciiBytesToChar(base64));
+            SecretKey secretKey = factory.generateSecret(passwordBasedKeySpec);
+            keyStore.setEntry(identifier.toExternalForm(), new KeyStore.SecretKeyEntry(secretKey), protectionParameter);
+            try (final OutputStream os = Files.newOutputStream(keyStorePath)) {
+                keyStore.store(os, keyStorePass);
+            } finally {
+                passwordBasedKeySpec.clearPassword();
+                SecretStoreUtil.clearBytes(secret);
```

Comment:
`secret` was already cleared when it was sent as an argument to `SecretStoreUtil#base64Encode(byte[])`, so clearing it again has no effect

It may be helpful to rename the methods that clear the passed `byte[]` to indicate that they do (e.g., `SecretStoreUtil#base64EncodeAndClear(byte[])`) so that it's easier to tell in calling code what has already been cleared.

Reply:
> secret was already cleared when it was sent as an argument

Agreed, but I just wanted to be explicit and not rely on a different methods side effect for this. 

> helpful to rename the methods

No strong objections to changing the name, but the reason I didn't add that to the name is the presumption that all things that all methods of the SecretStoreUtil has that side effect and thus every method would need that suffix making it abit redundant.